ports/raspberrypi/common-hal/pulseio/PulseIn.c

```diff
@@ -0,0 +1,243 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+
+#include <stdint.h>
+
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/pulseio/PulseIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "supervisor/shared/translate.h"
+#include "bindings/rp2pio/StateMachine.h"
+#include "common-hal/pulseio/PulseIn.h"
+
+pulseio_pulsein_obj_t* save_self;
+
+#define NO_PIN 0xff
+volatile bool last_level;
+volatile uint16_t level_count = 0;
+volatile uint16_t result = 0;
+volatile uint16_t buf_index = 0;
+
+uint16_t pulsein_program[] = {
+    0xe03f, //  0: set    x, 31
+    0x4001, //  1: in     pins, 1
+    0x0041, //  2: jmp    x--, 2
+    0x8060, //  3: push   iffull block
+    0xc020, //  4: irq    wait 0
+    0x0000, //  5: jmp    1
+};
+
+void common_hal_pulseio_pulsein_construct(pulseio_pulsein_obj_t* self,
+        const mcu_pin_obj_t* pin, uint16_t maxlen, bool idle_state) {
+
+    self->buffer = (uint16_t *) m_malloc(maxlen * sizeof(uint16_t), false);
+    if (self->buffer == NULL) {
+        mp_raise_msg_varg(&mp_type_MemoryError, translate("Failed to allocate RX buffer of %d bytes"), maxlen * sizeof(uint16_t));
+    }
+    self->pin = pin->number;
+    self->maxlen = maxlen;
+    self->idle_state = idle_state;
+    self->start = 0;
+    self->len = 0;
+    save_self = self;
+
+    // Set everything up.
+    rp2pio_statemachine_obj_t state_machine;
+
+    bool ok = rp2pio_statemachine_construct(&state_machine,
+        pulsein_program, sizeof(pulsein_program) / sizeof(pulsein_program[0]),
+        1000000 * 3,
+        NULL, 0,
+        NULL, 0,
+        pin, 1,
+        0,0,
+        NULL, 0,
+        NULL, 0,
+        1, 0,
+        1 << self->pin, false, true,
+        false, 8, false, // TX, unused
+        false,
+        true, 32, true, // RX auto-push every 32 bits
+        false); // claim pins
+    pio_sm_set_enabled(state_machine.pio,state_machine.state_machine, false);
+    self->state_machine.pio = state_machine.pio;
+    self->state_machine.state_machine = state_machine.state_machine;
+    self->state_machine.sm_config = state_machine.sm_config;
+    self->state_machine.offset = state_machine.offset;
+    if ( self->state_machine.pio == pio0 ) {
+       self->pio_interrupt = PIO0_IRQ_0;
+    } else {
+       self->pio_interrupt = PIO1_IRQ_0;
+    }
+    pio_sm_clear_fifos(self->state_machine.pio,self->state_machine.state_machine);
+    last_level = self->idle_state;
+    level_count = 0;
+    result = 0;
+    buf_index = 0;
+
+    pio_sm_set_in_pins(state_machine.pio,state_machine.state_machine,pin->number);
+    irq_set_exclusive_handler(self->pio_interrupt, common_hal_pulseio_pulsein_interrupt);
```

Comment:
`rotaryio` uses this interrupt too so Jeff added: https://github.com/adafruit/circuitpython/blob/main/ports/raspberrypi/common-hal/rp2pio/StateMachine.c#L772